### PR TITLE
Implement Bubble sort function plus tests

### PR DESCRIPTION
**Pull request by runnerdave**
Fixes #[42](https://github.com/anz-bank/go-course/issues/42) 

Review of colleague's PR #[227](https://github.com/anz-bank/go-course/pull/227) 

#### Changes proposed in this PR:

 - Implement bubble sort algorithm
 - Cover with 100% unit tests using tables

### CODE REVIEW

File: 02_bubble/runnerdave/main_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+		t.Errorf("Unexpected output in main()")
+	}
+}
+
+func TestSort(t *testing.T) {
+	tables := []struct {
+		x []int
+		n []int
+	}{
+		{[]int{1, 3, 2, 5}, []int{1, 2, 3, 5}},
```
Comment (anzboi) [inline 02_bubble/runnerdave/main_test.go:30]: Can you add tests for edge cases: empty and 1 element slices.
You should also test that the sort does not modify the original slice

File: 02_bubble/runnerdave/main.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprint(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(s []int) []int {
```
Comment (anzboi) [inline 02_bubble/runnerdave/main.go:15]: This function modifies `s`, it shouldn't. I realise you are passing in a copy in your tests, which is probably just as good, but we have to go with the lab instructions and require the function does the copy.

File: 02_bubble/runnerdave/main.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	len := len(s)
+	for i := 0; i < len-1; i++ {
+		for j := 0; j < len-i-1; j++ {
+			if s[j] > s[j+1] {
+				temp := s[j]
```
Comment (anzboi) [inline 02_bubble/runnerdave/main.go]: You can do this in one line using multiple assignment...
`a, b = b, a`
This doesn't lose one of the values as you might expect it to in other languages.

**Review by anzboi — CHANGES_REQUESTED** [review]
Looking good. Aside from the input modification issue this is just about done. Would also like to test a few edge cases.

File: 02_bubble/runnerdave/main_test.go (hunk 1 above)
Comment (runnerdave) [reply 02_bubble/runnerdave/main_test.go:30]: done

File: 02_bubble/runnerdave/main.go (hunk 2 above)
Comment (runnerdave) [reply 02_bubble/runnerdave/main.go:15]: good catch 😉

File: 02_bubble/runnerdave/main.go (hunk 3 above)
Comment (runnerdave) [reply 02_bubble/runnerdave/main.go]: this is very nice, thanks for the detailed review

**Comment by runnerdave** [comment]
@anzboi Hi Mate, any chance to finish this off? I implemented all the changes, kind regards

**Review by anzboi — APPROVED** [review]
LGTM

**Pull request merged**